File: 2019/04/rs/src/main.rs
use std::env;
use std::fs;
use std::time::Instant;

fn is_password_valid(password: &str, check2: bool) -> bool {
    let chars = password.chars().collect::<Vec<char>>();
    if (0..chars.len() - 1).all(|index| chars[index] <= chars[index + 1]) {
        let counts = password
            .chars()
            .collect::<std::collections::HashSet<char>>()
            .iter()
            .map(|c| password.matches(*c).count())
            .collect::<Vec<usize>>();
        return (!check2 || counts.contains(&2)) && counts.iter().any(|count| *count > 1);
    }
    false
}

fn get_valid_password_count(limits: &(u32, u32), check2: bool) -> usize {
    let (start, end) = limits;
    (*start..*end)
        .filter(|password| is_password_valid(&(password.to_string()), check2))
        .count()
}

fn solve(limits: &(u32, u32)) -> (usize, usize) {
    (
        get_valid_password_count(limits, false),
        get_valid_password_count(limits, true),
    )
}

fn get_input(file_path: &String) -> (u32, u32) {
    let split: Vec<String> = fs::read_to_string(file_path)
        .expect("Error reading input file!")
        .split("-")
        .map(|s| String::from(s))
        .collect();
    (split[0].parse().unwrap(), split[1].parse().unwrap())
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:7}", now.elapsed().as_secs_f32());
}

File: 2019/04/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>

typedef struct
{
    int start, end;
} Input;
typedef struct
{
    int part1;
    int part2;
} Results;

int isPasswordValid(int password, int check2)
{
    int previous = 9, current, counts[10], index = 10, hasTwo = !check2, hasGreaterThanOne = 0;
    for (index = 0; index < 10; index++)
        counts[index] = 0;
    for (; password && index; --index, password /= 10)
    {
        current = password % 10;
        if (previous < current)
            return 0;
        counts[current]++;
        previous = current;
    }
    for (index = 0; index < 10; index++)
    {
        hasTwo |= counts[index] == 2;
        hasGreaterThanOne |= counts[index] > 1;
    }
    return hasTwo && hasGreaterThanOne;
}

int getValidPasswordCount(Input limits, int check2)
{
    int count = 0;
    for (int password = limits.start; password < limits.end; password++)
        count += isPasswordValid(password, check2);
    return count;
}

Results solve(Input input)
{
    return (Results){getValidPasswordCount(input, 0), getValidPasswordCount(input, 1)};
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    fseek(file, 0, SEEK_END);
    long length = ftell(file);
    rewind(file);
    char contents[length + 1];
    fread(contents, length, 1, file);
    int start, end;
    sscanf(contents, "%d-%d", &start, &end);
    fclose(file);
    return (Input){
        start,
        end};
}

void freeInput(Input input)
{
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}